File: tests/test_history_file.c
/*
 * test_history_file.c - Unit tests for history file functionality
 * Tests newline escaping/unescaping and dynamic buffer resizing
 */

#define TEST_BUILD 1

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <assert.h>
#include "../src/history_file.h"

// Test framework colors
#define COLOR_RESET "\033[0m"
#define COLOR_GREEN "\033[32m"
#define COLOR_RED "\033[31m"
#define COLOR_YELLOW "\033[33m"
#define COLOR_CYAN "\033[36m"

// Test counters
static int tests_run = 0;
static int tests_passed = 0;
static int tests_failed = 0;

// Test utilities
static void print_test_result(const char *test_name, int passed) {
    tests_run++;
    if (passed) {
        tests_passed++;
        printf(COLOR_GREEN "✓ PASS" COLOR_RESET " %s\n", test_name);
    } else {
        tests_failed++;
        printf(COLOR_RED "✗ FAIL" COLOR_RESET " %s\n", test_name);
    }
}

static void print_summary(void) {
    printf("\n" COLOR_CYAN "Test Summary:" COLOR_RESET "\n");
    printf("Tests run: %d\n", tests_run);
    printf(COLOR_GREEN "Tests passed: %d\n" COLOR_RESET, tests_passed);
    if (tests_failed > 0) {
        printf(COLOR_RED "Tests failed: %d\n" COLOR_RESET, tests_failed);
    } else {
        printf(COLOR_GREEN "All tests passed!\n" COLOR_RESET);
    }
}

#define ASSERT(cond) do { \
    if (!(cond)) { \
        printf("%s[FAIL] %s:%d: Assertion failed: %s%s\n", COLOR_RED, __FILE__, __LINE__, #cond, COLOR_RESET); \
        return 0; \
    } \
} while(0)

static int test_escape_newlines_basic(void) {
    printf("\n%s[TEST] test_escape_newlines_basic%s\n", COLOR_CYAN, COLOR_RESET);

    // Test basic escaping
    char *escaped = escape_newlines("hello\nworld");
    ASSERT(escaped != NULL);
    ASSERT(strcmp(escaped, "hello\\nworld") == 0);
    free(escaped);

    // Test multiple newlines
    escaped = escape_newlines("line1\nline2\nline3");
    ASSERT(escaped != NULL);
    ASSERT(strcmp(escaped, "line1\\nline2\\nline3") == 0);
    free(escaped);

    // Test no newlines
    escaped = escape_newlines("no newlines here");
    ASSERT(escaped != NULL);
    ASSERT(strcmp(escaped, "no newlines here") == 0);
    free(escaped);

    // Test empty string
    escaped = escape_newlines("");
    ASSERT(escaped != NULL);
    ASSERT(strcmp(escaped, "") == 0);
    free(escaped);

    // Test NULL input
    escaped = escape_newlines(NULL);
    ASSERT(escaped == NULL);

    return 1;
}

static int test_unescape_newlines_basic(void) {
    printf("\n%s[TEST] test_unescape_newlines_basic%s\n", COLOR_CYAN, COLOR_RESET);

    // Test basic unescaping
    char *unescaped = unescape_newlines("hello\\nworld");
    ASSERT(unescaped != NULL);
    ASSERT(strcmp(unescaped, "hello\nworld") == 0);
    free(unescaped);

    // Test multiple escaped sequences
    unescaped = unescape_newlines("line1\\nline2\\nline3");
    ASSERT(unescaped != NULL);
    ASSERT(strcmp(unescaped, "line1\nline2\nline3") == 0);
    free(unescaped);

    // Test no escape sequences
    unescaped = unescape_newlines("no escapes here");
    ASSERT(unescaped != NULL);
    ASSERT(strcmp(unescaped, "no escapes here") == 0);
    free(unescaped);

    // Test empty string
    unescaped = unescape_newlines("");
    ASSERT(unescaped != NULL);
    ASSERT(strcmp(unescaped, "") == 0);
    free(unescaped);

    // Test NULL input
    unescaped = unescape_newlines(NULL);
    ASSERT(unescaped == NULL);

    return 1;
}

static int test_escape_unescape_roundtrip(void) {
    printf("\n%s[TEST] test_escape_unescape_roundtrip%s\n", COLOR_CYAN, COLOR_RESET);

    const char *test_cases[] = {
        "single line",
        "line1\nline2",
        "line1\nline2\nline3",
        "text with\nmultiple\nnewlines\nin it",
        "",
        "\n",
        "\n\n\n",
        "text\n",
        "\ntext",
        "text\n\ntext",
        NULL
    };

    for (int i = 0; test_cases[i] != NULL; i++) {
        char *escaped = escape_newlines(test_cases[i]);
        ASSERT(escaped != NULL);

        char *unescaped = unescape_newlines(escaped);
        ASSERT(unescaped != NULL);

        ASSERT(strcmp(unescaped, test_cases[i]) == 0);

        free(escaped);
        free(unescaped);
    }

    return 1;
}

static int test_history_file_append_with_newlines(void) {
    printf("\n%s[TEST] test_history_file_append_with_newlines%s\n", COLOR_CYAN, COLOR_RESET);

    // Create temporary file for testing
    char temp_path[] = "/tmp/test_history_XXXXXX";
    int fd = mkstemp(temp_path);
    ASSERT(fd != -1);
    close(fd);

    HistoryFile *hf = history_file_open(temp_path);
    ASSERT(hf != NULL);

    // Test appending text with newlines
    const char *text_with_newlines = "line1\nline2\nline3";
    int result = history_file_append(hf, text_with_newlines);
    ASSERT(result == 0);

    // Close and reopen to verify persistence
    history_file_close(hf);
    hf = history_file_open(temp_path);
    ASSERT(hf != NULL);

    // Load and verify the escaped content
    int count = 0;
    char **lines = history_file_load_recent(hf, 10, &count);
    ASSERT(lines != NULL);
    ASSERT(count == 1);
    ASSERT(strcmp(lines[0], text_with_newlines) == 0);

    // Cleanup
    for (int i = 0; i < count; i++) {
        free(lines[i]);
    }
    free(lines);
    history_file_close(hf);
    unlink(temp_path);

    return 1;
}

static int test_history_file_load_recent_with_escaped_newlines(void) {
    printf("\n%s[TEST] test_history_file_load_recent_with_escaped_newlines%s\n", COLOR_CYAN, COLOR_RESET);

    // Create temporary file with escaped newlines
    char temp_path[] = "/tmp/test_history_escaped_XXXXXX";
    int fd = mkstemp(temp_path);
    ASSERT(fd != -1);

    // Write escaped content directly
    FILE *fp = fdopen(fd, "w");
    ASSERT(fp != NULL);
    fprintf(fp, "line1\\nline2\\nline3\n");
    fprintf(fp, "single line\n");
    fclose(fp);

    HistoryFile *hf = history_file_open(temp_path);
    ASSERT(hf != NULL);

    // Load and verify unescaped content
    int count = 0;
    char **lines = history_file_load_recent(hf, 10, &count);
    ASSERT(lines != NULL);
    ASSERT(count == 2);
    ASSERT(strcmp(lines[0], "line1\nline2\nline3") == 0);
    ASSERT(strcmp(lines[1], "single line") == 0);

    // Cleanup
    for (int i = 0; i < count; i++) {
        free(lines[i]);
    }
    free(lines);
    history_file_close(hf);
    unlink(temp_path);

    return 1;
}

static int test_history_file_edge_cases(void) {
    printf("\n%s[TEST] test_history_file_edge_cases%s\n", COLOR_CYAN, COLOR_RESET);

    char temp_path[] = "/tmp/test_history_edge_XXXXXX";
    int fd = mkstemp(temp_path);
    ASSERT(fd != -1);
    close(fd);

    HistoryFile *hf = history_file_open(temp_path);
    ASSERT(hf != NULL);

    // Test empty string (should be skipped)
    int result = history_file_append(hf, "");
    ASSERT(result == 0);  // Should succeed but not write anything

    // Test NULL text
    result = history_file_append(hf, NULL);
    ASSERT(result == 0);  // Should succeed but not write anything

    // Test valid text
    result = history_file_append(hf, "valid text");
    ASSERT(result == 0);

    // Verify only valid text was written
    int count = 0;
    char **lines = history_file_load_recent(hf, 10, &count);
    ASSERT(lines != NULL);
    ASSERT(count == 1);
    ASSERT(strcmp(lines[0], "valid text") == 0);

    // Cleanup
    for (int i = 0; i < count; i++) {
        free(lines[i]);
    }
    free(lines);
    history_file_close(hf);
    unlink(temp_path);

    return 1;
}

int main(void) {
    printf("%sRunning History File Tests%s\n", COLOR_CYAN, COLOR_RESET);

    print_test_result("test_escape_newlines_basic", test_escape_newlines_basic());
    print_test_result("test_unescape_newlines_basic", test_unescape_newlines_basic());
    print_test_result("test_escape_unescape_roundtrip", test_escape_unescape_roundtrip());
    print_test_result("test_history_file_append_with_newlines", test_history_file_append_with_newlines());
    print_test_result("test_history_file_load_recent_with_escaped_newlines", test_history_file_load_recent_with_escaped_newlines());
    print_test_result("test_history_file_edge_cases", test_history_file_edge_cases());

    print_summary();

    return (tests_failed == 0) ? 0 : 1;
}

File: zig/tests/test_history_file.zig
//! tests/test_history_file.zig — Zig port of tests/test_history_file.c
//!
//! Tests the HistoryFile API: newline escaping/unescaping, append,
//! loadRecent, limit handling, and edge cases.

const std = @import("std");
const history_file = @import("../history_file.zig");

const HistoryFile = history_file.HistoryFile;

// ---------------------------------------------------------------------------
// escapeNewlines
// ---------------------------------------------------------------------------

test "history_file: escape basic newlines" {
    const alloc = std.testing.allocator;

    const escaped = try history_file.escapeNewlines(alloc, "hello\nworld");
    defer alloc.free(escaped);

    try std.testing.expectEqualStrings("hello\\nworld", escaped);
}

test "history_file: escape multiple newlines" {
    const alloc = std.testing.allocator;

    const escaped = try history_file.escapeNewlines(alloc, "line1\nline2\nline3");
    defer alloc.free(escaped);

    try std.testing.expectEqualStrings("line1\\nline2\\nline3", escaped);
}

test "history_file: escape no-op when no newlines" {
    const alloc = std.testing.allocator;

    const escaped = try history_file.escapeNewlines(alloc, "no newlines here");
    defer alloc.free(escaped);

    try std.testing.expectEqualStrings("no newlines here", escaped);
}

test "history_file: escape empty string" {
    const alloc = std.testing.allocator;

    const escaped = try history_file.escapeNewlines(alloc, "");
    defer alloc.free(escaped);

    try std.testing.expectEqualStrings("", escaped);
}

test "history_file: escaped output contains no literal newlines" {
    const alloc = std.testing.allocator;

    const escaped = try history_file.escapeNewlines(alloc, "a\nb\nc");
    defer alloc.free(escaped);

    try std.testing.expect(std.mem.indexOf(u8, escaped, "\n") == null);
}

// ---------------------------------------------------------------------------
// unescapeNewlines
// ---------------------------------------------------------------------------

test "history_file: unescape basic escape sequence" {
    const alloc = std.testing.allocator;

    const unescaped = try history_file.unescapeNewlines(alloc, "hello\\nworld");
    defer alloc.free(unescaped);

    try std.testing.expectEqualStrings("hello\nworld", unescaped);
}

test "history_file: unescape multiple sequences" {
    const alloc = std.testing.allocator;

    const unescaped = try history_file.unescapeNewlines(alloc, "line1\\nline2\\nline3");
    defer alloc.free(unescaped);

    try std.testing.expectEqualStrings("line1\nline2\nline3", unescaped);
}

test "history_file: unescape no-op when no escape sequences" {
    const alloc = std.testing.allocator;

    const result = try history_file.unescapeNewlines(alloc, "no escapes here");
    defer alloc.free(result);

    try std.testing.expectEqualStrings("no escapes here", result);
}

test "history_file: unescape empty string" {
    const alloc = std.testing.allocator;

    const result = try history_file.unescapeNewlines(alloc, "");
    defer alloc.free(result);

    try std.testing.expectEqualStrings("", result);
}

// ---------------------------------------------------------------------------
// Round-trip: escape → unescape
// ---------------------------------------------------------------------------

test "history_file: escape/unescape round-trip single newline" {
    const alloc = std.testing.allocator;

    const original = "line1\nline2";
    const escaped = try history_file.escapeNewlines(alloc, original);
    defer alloc.free(escaped);
    const restored = try history_file.unescapeNewlines(alloc, escaped);
    defer alloc.free(restored);

    try std.testing.expectEqualStrings(original, restored);
}

test "history_file: escape/unescape round-trip multiple values" {
    const alloc = std.testing.allocator;

    const cases = [_][]const u8{
        "single line",
        "line1\nline2",
        "line1\nline2\nline3",
        "text with\nmultiple\nnewlines\nin it",
        "",
        "\n",
        "\n\n\n",
        "text\n",
        "\ntext",
        "text\n\ntext",
    };

    for (cases) |original| {
        const escaped = try history_file.escapeNewlines(alloc, original);
        defer alloc.free(escaped);
        const restored = try history_file.unescapeNewlines(alloc, escaped);
        defer alloc.free(restored);
        try std.testing.expectEqualStrings(original, restored);
    }
}

// ---------------------------------------------------------------------------
// HistoryFile open, append, loadRecent
// ---------------------------------------------------------------------------

test "history_file: append and loadRecent basic" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "hist.txt" });
    defer alloc.free(path);

    var hf = try HistoryFile.open(alloc, path);
    defer hf.close();

    try hf.append("first entry");
    try hf.append("second entry");
    try hf.append("third entry");

    const entries = try hf.loadRecent(10);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 3), entries.len);
    try std.testing.expectEqualStrings("first entry", entries[0]);
    try std.testing.expectEqualStrings("second entry", entries[1]);
    try std.testing.expectEqualStrings("third entry", entries[2]);
}

test "history_file: append with newlines preserves multi-line content" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "multiline.txt" });
    defer alloc.free(path);

    var hf = try HistoryFile.open(alloc, path);
    defer hf.close();

    const text_with_newlines = "line1\nline2\nline3";
    try hf.append(text_with_newlines);

    const entries = try hf.loadRecent(10);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 1), entries.len);
    try std.testing.expectEqualStrings(text_with_newlines, entries[0]);
}

test "history_file: loadRecent from file with pre-escaped content" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "preescaped.txt" });
    defer alloc.free(path);

    // Write escaped content directly to the file
    {
        const file = try std.fs.cwd().createFile(path, .{});
        defer file.close();
        try file.writeAll("line1\\nline2\\nline3\n");
        try file.writeAll("single line\n");
    }

    var hf = try HistoryFile.open(alloc, path);
    defer hf.close();

    const entries = try hf.loadRecent(10);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 2), entries.len);
    try std.testing.expectEqualStrings("line1\nline2\nline3", entries[0]);
    try std.testing.expectEqualStrings("single line", entries[1]);
}

// ---------------------------------------------------------------------------
// loadRecent limit
// ---------------------------------------------------------------------------

test "history_file: loadRecent respects limit" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "limit.txt" });
    defer alloc.free(path);

    var hf = try HistoryFile.open(alloc, path);
    defer hf.close();

    var i: usize = 0;
    while (i < 5) : (i += 1) {
        const line = try std.fmt.allocPrint(alloc, "entry {d}", .{i});
        defer alloc.free(line);
        try hf.append(line);
    }

    const entries = try hf.loadRecent(3);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 3), entries.len);
    try std.testing.expectEqualStrings("entry 2", entries[0]);
    try std.testing.expectEqualStrings("entry 3", entries[1]);
    try std.testing.expectEqualStrings("entry 4", entries[2]);
}

test "history_file: loadRecent limit 0 returns empty slice" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "zero.txt" });
    defer alloc.free(path);

    var hf = try HistoryFile.open(alloc, path);
    defer hf.close();

    try hf.append("something");

    const entries = try hf.loadRecent(0);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 0), entries.len);
}

// ---------------------------------------------------------------------------
// Edge cases
// ---------------------------------------------------------------------------

test "history_file: appending empty string is silently ignored" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "empty.txt" });
    defer alloc.free(path);

    var hf = try HistoryFile.open(alloc, path);
    defer hf.close();

    try hf.append(""); // should succeed but write nothing

    const entries = try hf.loadRecent(10);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 0), entries.len);
}

test "history_file: mix of empty and non-empty appends" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "mixed.txt" });
    defer alloc.free(path);

    var hf = try HistoryFile.open(alloc, path);
    defer hf.close();

    try hf.append("valid text");
    try hf.append(""); // should be skipped
    try hf.append("also valid");

    const entries = try hf.loadRecent(10);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 2), entries.len);
    try std.testing.expectEqualStrings("valid text", entries[0]);
    try std.testing.expectEqualStrings("also valid", entries[1]);
}

test "history_file: loadRecent on non-existent file returns empty slice" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    // Create a HistoryFile pointing at an existing path, then remove the file
    // and try to loadRecent — must return empty, not crash.
    const path = try std.fs.path.join(alloc, &.{ dir, "ghost.txt" });
    defer alloc.free(path);

    var hf = try HistoryFile.open(alloc, path);
    defer hf.close();

    // Remove the underlying file while the handle is still open
    std.fs.cwd().deleteFile(path) catch {};

    const entries = try hf.loadRecent(10);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 0), entries.len);
}

// ---------------------------------------------------------------------------
// Persistence: close and reopen
// ---------------------------------------------------------------------------

test "history_file: entries persist across close/reopen" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "persist.txt" });
    defer alloc.free(path);

    {
        var hf = try HistoryFile.open(alloc, path);
        defer hf.close();
        try hf.append("persistent entry");
    }

    // Re-open and verify
    var hf2 = try HistoryFile.open(alloc, path);
    defer hf2.close();

    const entries = try hf2.loadRecent(10);
    defer hf2.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 1), entries.len);
    try std.testing.expectEqualStrings("persistent entry", entries[0]);
}
